### PR TITLE
chore: reduce CI complexity. use nx cloud agents

### PR DESCRIPTION
**Pull request by hoebbelsB**
not sure about the `parallel` usage, but everything else should be just fine. Spent some time on tweaking the agents setup. We of course could opt-in to use those xtra-large images ... but i'm kind of hesitating to use so much power for our little lib :D

### CODE REVIEW

**Comment by hoebbelsB** [comment]
@eneajaho do u think it's working with the current `--parallel` setting? We also run e2e and component tests with that. Not sure what issues can be caused because of it

**Comment by LayZeeDK** [comment]
> @eneajaho do u think it's working with the current `--parallel` setting? We also run e2e and component tests with that. Not sure what issues can be caused because of it

@hoebbelsB 

When using Nx Agents, it's recommended not to have a `--parallel` setting [as seen in the docs](https://nx.dev/ci/features/split-e2e-tasks#configure-automated-task-splitting-on-ci) and in [the workflow generator](https://github.com/nrwl/nx/blob/18e5535940a0aff86dc845f3714b07b94a0db8d8/packages/workspace/src/generators/ci-workflow/files/github/.github/workflows/__workflowFileName__.yml__tmpl__). It's also recommended to [use Nx Atomizer](https://nx.dev/ci/features/split-e2e-tasks), that is the `e2e-ci` tasks to distribute individual end-to-end tests. In this case, `e2e-ci` should also be passed to `--stop-agents-after`

File: .github/workflows/build-and-test.yml [hunk 1]
```diff
@@ -81,53 +44,26 @@ jobs:
...
-            fi
-          done
-
-          exit 0
-
-      - name: Stop Nx Cloud DTE agents
-        if: ${{ always() }}
-        run: yarn nx-cloud stop-all-agents
+          npx nx-cloud record -- npx nx format:check
+          npx nx affected -t lint build test component-test e2e --parallel=4 --exclude=docs
```
Comment (LayZeeDK) [inline .github/workflows/build-and-test.yml:66]: suggestion: Remove the `--parallel` option as distribution is optimized by Nx Agents based on the activated changeset.

```suggestion
          npx nx affected -t lint build test component-test e2e --exclude=docs
```

Comment (LayZeeDK) [inline .github/workflows/build-and-test.yml:66]: suggestion: Try using the `e2e-ci` tasks. [Configuration docs for the Nx Cypress plugin](https://nx.dev/nx-api/cypress#nxcypress-configuration).

Specify `e2e-ci` both for `nx cloud start-ci-run --stop-agents-after` and `nx affected -t`.

```suggestion
          npx nx affected -t lint build test component-test e2e-ci --parallel=4 --exclude=docs
```

File: .github/workflows/build-and-test.yml [hunk 2]
```diff
@@ -81,53 +44,26 @@ jobs:
...
           fetch-depth: 0
 
-      - name: Set up dependencies
-        uses: ./.github/actions/setup
+      - name: Derive appropriate SHAs for base and head for `nx affected` commands
+        uses: nrwl/nx-set-shas@v4
 
       - name: Initialize the Nx Cloud distributed CI run
-        run: yarn nx-cloud start-ci-run
+        run: npx nx-cloud start-ci-run --distribute-on="./nx/workflows/dynamic-changesets.yml" --stop-agents-after="e2e"
```
Comment (LayZeeDK) [inline .github/workflows/build-and-test.yml:58]: suggestion: Try using the `e2e-ci` tasks. [Configuration docs for the Nx Cypress plugin](https://nx.dev/nx-api/cypress#nxcypress-configuration).

Specify `e2e-ci` both for `nx cloud start-ci-run --stop-agents-after` and `nx affected -t`.

```suggestion
        run: npx nx-cloud start-ci-run --distribute-on="./nx/workflows/dynamic-changesets.yml" --stop-agents-after="e2e-ci"
```

**Review by LayZeeDK — COMMENTED** [review]
Suggest changes.

File: .github/workflows/build-and-test.yml (hunk 1 above)
Comment (hoebbelsB) [reply .github/workflows/build-and-test.yml]: thanks!

File: .github/workflows/build-and-test.yml (hunk 2 above)
Comment (hoebbelsB) [reply .github/workflows/build-and-test.yml]: @LayZeeDK tried, didn't work. I don't want to mess around with it any longer :D if you have any other suggestion, i'm going to revert the e2e-ci changes

File: .github/workflows/build-and-test.yml (hunk 1 above)
Comment (hoebbelsB) [reply .github/workflows/build-and-test.yml]: @LayZeeDK tried, didn't work. I don't want to mess around with it any longer :D if you have any other suggestion, i'm going to revert the e2e-ci changes

Comment (LayZeeDK) [reply .github/workflows/build-and-test.yml]: @eneajaho

Have you tried ISR with Cypress and Nx Atomizer (`e2e-ci` tasks)?

Comment (LayZeeDK) [reply .github/workflows/build-and-test.yml]: Oh, it's also SSR without ISR

https://cloud.nx.app/runs/16A7H6itJx/task/ssr%3Ae2e-ci--cypress%2Fe2e%2Fapp.cy.ts

Comment (hoebbelsB) [reply .github/workflows/build-and-test.yml]: @LayZeeDK I don't see any evidence about `--parallel` is a bad idea for cloud agents. In fact, tried it on my private projects.

image

You can see that the agent is running 4 tasks in parallel, as I am running it with `--parallel=4`. Imo it should be better to have them run things in parallel, wdyt?

Comment (LayZeeDK) [reply .github/workflows/build-and-test.yml]: Are you using Nx Agents in your private project?

```shell
nx-cloud start-ci-run --distribute-on="./nx/workflows/dynamic-changesets.yml"
```

Comment (hoebbelsB) [reply .github/workflows/build-and-test.yml]: @LayZeeDK yes!

**Pull request merged**